SMLHelper/Utility/ItemStorageHelper.cs

```diff
@@ -0,0 +1,124 @@
+﻿namespace SMLHelper.V2.Utility
+{
+	using System.Collections.Generic;
+
+	public class ItemStorageHelper : IStorageHelper
+	{
+		private readonly Dictionary<IItemsContainer, Dictionary<Vector2int, bool>> HasRoomCacheCollection = new Dictionary<IItemsContainer, Dictionary<Vector2int, bool>>();
+
+		public static readonly ItemStorageHelper Helper = new ItemStorageHelper();
+
+		private readonly IEnumerable<Vector2int> SmallerThan3x3 = new Vector2int[]
+		{
+			new Vector2int(3,2),
+			new Vector2int(3,1),
+			new Vector2int(2,3),
+			new Vector2int(2,2),
+			new Vector2int(2,1),
+			new Vector2int(1,2),
+			new Vector2int(1,1)
+		};
+
+		private readonly IEnumerable<Vector2int> SmallerThan2x3 = new Vector2int[]
+		{
+			new Vector2int(2,2),
+			new Vector2int(2,1),
+			new Vector2int(1,2),
+			new Vector2int(1,1)
+		};
+
+		private readonly IEnumerable<Vector2int> SmallerThan2x2 = new Vector2int[]
+		{
+			new Vector2int(2,1),
+			new Vector2int(1,2),
+			new Vector2int(1,1)
+		};
+
+		private readonly IEnumerable<Vector2int> Just1x1 = new Vector2int[]
+		{
+			new Vector2int(1,1)
+		};
+
+		// Used to quickly cache common sizes if a large item is found to fit a container
+		private IEnumerable<Vector2int> CommonSmallerSizes(Vector2int original)
+		{
+			if (original.x == 3 && original.y == 3) // Size of Mobile Vehicle Bay
+			{
+				return this.SmallerThan3x3;
+			}
+
+			if (original.x == 2 && original.y == 3) // Size of Seaglide
+			{
+				return this.SmallerThan2x3;
+			}
+
+			if (original.x == 2 && original.y == 2) // Size of common game items
+			{
+				return this.SmallerThan2x2;
+			}
+
+			return this.Just1x1;
+		}
+
+		private void CacheNewContainer(IItemsContainer container)
+		{
+			this.HasRoomCacheCollection.Add(container, new Dictionary<Vector2int, bool>());
+		}
+
+		internal void RecacheContainer(IItemsContainer container)
+		{
+			// Items in storage have changed. Clear the related cache.
+			if (this.HasRoomCacheCollection.TryGetValue(container, out Dictionary<Vector2int, bool> cache))
+			{
+				cache.Clear();
+			}
+
+			else
+			{
+				// This is a new container we haven't seen before, save it to the cache collection.
+				this.CacheNewContainer(container);
+			}
+		}
+
+		internal void CacheNewHasRoomData(IItemsContainer container, Vector2int itemSize, bool hasRoom)
+		{
+			this.HasRoomCacheCollection[container][itemSize] = hasRoom;
+
+			// If item fits and is larger than 1x1, cache common sizes as true
+			if (hasRoom && (itemSize.x > 1 || itemSize.y > 1))
+			{
+				foreach (Vector2int size in this.CommonSmallerSizes(itemSize))
+				{
+					this.HasRoomCacheCollection[container][size] = true;
+				}
+			}
+		}
+
+		public bool TryGetCachedHasRoom(IItemsContainer container, int width, int height, ref bool cachedResult)
```

Comment:
@DingoDjango said
> If a mod calls `container.TryGetCachedHasRoom` and there is no cache, should we run our own HasRoom calculation?
> Right now we are only caching when called from Harmony (because the game runs the calculation). If a mod calls `TryGet` it will return false without caching anything.
That is indeed the original behavior.  
This whole patch was created under the expectation that the only time these methods are being called is because there is a game event requesting it.  
Now we open the possibility of mods to start checking these methods outside of the normal game loops.

At this point, a decision should be made.  
In my opinion, because the mod already has access to the `IItemsContainer` it wants to deal with,  
when it gets a `false` return value from `TryGetCachedHasRoom` it can simply go straight into the original `HasRoomFor`, with the benefit of the patches here caching the value.

It seems bit redundant, but when the data just isn't available, there isn't much else to do.

Reply:
So just to be clear, are we going to fall back to calling the original `HasRoomFor` or do we leave that up to others mods?
My thinking was make a `TryGetCachedHasRoomInternal` for the Harmony patch, and other `TryGet` methods will be simplified and default to calling `ItemsContainer.HasRoom` if they don't get a result right away.

Reply:
If the `IItemsContainer` in question is not `ItemsContainer` then we still return false, of course.

Reply:
It shouldn't matter what actually call is behind the `IItemsContainer` interface.  
This interface has all the methods we ever need to interact with.

As for what a mod will do when it gets a `false` out of the `TryGetCachedHasRoom` that seems best left to the external caller to decide.

Reply:
Addressed as we have shifted to only addressing `ItemsContainer`